File: WebClient SDK/src/PnrScripts/UI/PnrActUIExtension.ts
import { CardUIExtension, ICardUIExtensionContext } from 'tessa/ui/cards';
import { Guid } from 'tessa/platform';
import { CardStoreMode } from 'tessa/cards';
import { DateTimeViewModel } from 'tessa/ui/cards/controls';
import { PnrCardTypes } from '../Shared/PnrCardTypes';
import { PnrActTypes } from '../Shared/PnrActTypes';
import Moment from 'moment';

export class PnrActUIExtension extends CardUIExtension {
  public initialized(context: ICardUIExtensionContext) {
    if (!Guid.equals(context.card.typeId, PnrCardTypes.PnrActTypeID)) {
      return;
    }
    let pnrActs = context.card.sections.tryGet("PnrActs");
    if (!pnrActs) return;

    // Создание карточки
    if (context.card.storeMode == CardStoreMode.Insert)
    {
        const projectDateControl = context.model.controls.get('ProjectDate') as DateTimeViewModel;
        if (!projectDateControl) return;
        // Заполнить поле Дата проекта текущей датой
        projectDateControl.selectedDate = Moment(Date.now());
    }

    pnrActs.fields.fieldChanged.add(async (e) => {
      // Тип акта: обязательность заполнения полей
      if (e.fieldName == "TypeID" && e.fieldValue != null)
      {
         this.ChangeControlsActiveValidation(e.fieldValue, context);
      }
    });
  }

  /** валидация полей, зависимых от Типа акта */
  private ChangeControlsActiveValidation(typeID, context: ICardUIExtensionContext)
  {
    // Стадия реализации
    const implementationStage = context.model.controls.get('ImplementationStage');
    if (implementationStage) {
      implementationStage.isRequired = typeID == PnrActTypes.PnrAcceptanceCertificateTypeID;
      implementationStage.notifyUpdateValidation();
    }
  }
}

File: Extensions/Tessa.Extensions.Client/UI/PnrActUIExtension.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Tessa.Cards;
using Tessa.Extensions.Shared.PnrCards;
using Tessa.UI.Cards;
using Tessa.UI.Cards.Controls;

namespace Tessa.Extensions.Client.UI
{
    public sealed class PnrActUIExtension : CardUIExtension
    {
        /// <summary>
        /// Валидация контролов.
        /// </summary>
        private void ChangeControlsActiveValidation(Guid typeID, ICardModel cardModel)
        {
            // Стадия реализации
            if (cardModel.Controls.TryGet("ImplementationStage", out IControlViewModel implementationStage))
            {
                implementationStage.IsRequired = typeID == PnrActTypes.PnrAcceptanceCertificateTypeID;
                implementationStage.NotifyUpdateValidation();
            }
        }
        public override Task Initialized(ICardUIExtensionContext context)
        {
            Card card = context.Card;
            ICardModel cardModel = context.Model;

            // Создание карточки
            if (card.StoreMode == CardStoreMode.Insert)
            {
                // Заполнить поле Дата проекта текущей датой
                if (cardModel.Controls.TryGet("ProjectDate", out IControlViewModel projectDateControl))
                {
                    ((DateTimeViewModel)projectDateControl).SelectedDate = DateTime.Now;
                }
            }

            context.Card.Sections["PnrActs"].FieldChanged += (s, e) =>
            {
                // Тип акта: обязательность заполнения полей
                if (e.FieldName == "TypeID" && e.FieldValue != null)
                {
                    ChangeControlsActiveValidation((Guid)e.FieldValue, cardModel);
                }
            };

            return base.Initialized(context);
        }
    }
}
